// @insforge/shared-schemas/dist/ai-chat.schema.js
import { z } from 'zod';
// ============= Chat Types =============
export const chatMessageSchema = z.object({
    role: z.enum(['user', 'assistant', 'system']),
    content: z.string(),
});
export const chatOptionsSchema = z.object({
    model: z.string(),
    temperature: z.number().min(0).max(2).optional(),
    maxTokens: z.number().positive().optional(),
    topP: z.number().min(0).max(1).optional(),
    systemPrompt: z.string().optional(),
});
export const chatRequestSchema = z.object({
    model: z.string(),
    messages: z.array(chatMessageSchema).optional(),
    stream: z.boolean().optional(),
    temperature: z.number().min(0).max(2).optional(),
    maxTokens: z.number().positive().optional(),
    topP: z.number().min(0).max(1).optional(),
    systemPrompt: z.string().optional(),
});
export const chatCompletionResponseSchema = z.object({
    success: z.boolean(),
    content: z.string(),
    metadata: z
        .object({
        model: z.string(),
        usage: z
            .object({
            promptTokens: z.number().optional(),
            completionTokens: z.number().optional(),
            totalTokens: z.number().optional(),
        })
            .optional(),
    })
        .optional(),
});
