// @insforge/shared-schemas/dist/ai-image.schema.js
import { z } from 'zod';
// ============= Image Types =============
export const imageGenerationOptionsSchema = z.object({
    model: z.string(),
    prompt: z.string(),
    // OpenRouter's image generation API currently only supports model, prompt, and modalities
});
export const generatedImageSchema = z.object({
    url: z.string().optional(),
    imageData: z.string().optional(),
    revisedPrompt: z.string().optional(),
});
// OpenRouter image generation response format
// According to OpenRouter docs, they add an 'images' field to the assistant message
export const openRouterImageMessageSchema = z.object({
    type: z.literal('image_url'),
    image_url: z.object({
        url: z.string(), // Can be a direct URL or data:image base64 URL
    }),
});
// Image Generation response format
export const imageGenerationResponseSchema = z.object({
    text: z.string().optional(), // Text content from multimodal models
    images: z.array(openRouterImageMessageSchema),
    metadata: z
        .object({
        model: z.string(),
        revisedPrompt: z.string().optional(),
        usage: z
            .object({
            promptTokens: z.number().optional(),
            completionTokens: z.number().optional(),
            totalTokens: z.number().optional(),
        })
            .optional(),
    })
        .optional(),
});
